### PR TITLE
[vividus-plugin-json] Add step to validate JSON against schema

### PR DESCRIPTION
**Pull request by IBrun**

### CODE REVIEW

File: docs/modules/plugins/pages/plugin-json.adoc [hunk 1]
```diff
@@ -500,3 +502,59 @@ Then number of JSON elements from `
...
+=== Validate JSON document
+
+Validates JSON against {json-schema}.
+
+[source,gherkin]
+----
+Then JSON `$json` is valid against schema `$schema`
+----
+
+* `$json` - The JSON.
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-json.adoc:515]:
```suggestion
* `$json` - The JSON to validate.
```

File: vividus-plugin-json/src/main/java/org/vividus/json/steps/JsonSteps.java [hunk 2]
```diff
@@ -485,6 +490,25 @@ public void assertNumberOfJsonElements(String json, String jsonPath, ComparisonR
...
 
+    /**
+     * Validates json against json schema
+     * @param json JSON
+     * @param schema JSON schema
+     */
+    @Then("JSON `$json` is valid against schema `$schema`")
+    public void validateJsonAgainstSchema(String json, String schema)
+    {
+        JsonSchemaFactory factory = JsonSchemaFactory.getInstance(SpecVersion.VersionFlag.V202012);
```
Comment (valfirst) [inline vividus-plugin-json/src/main/java/org/vividus/json/steps/JsonSteps.java:501]: I guess, it makes sense to version of the schema specified in it, otherwise we should default to 2020-12

File: vividus-plugin-json/src/main/java/org/vividus/json/steps/JsonSteps.java [hunk 3]
```diff
@@ -485,6 +490,25 @@ public void assertNumberOfJsonElements(String json, String jsonPath, ComparisonR
...
+        JsonSchema jsonSchema = factory.getSchema(schema);
+        JsonNode jsonNode = jsonUtils.readTree(json);
+        Set<ValidationMessage> errors = jsonSchema.validate(jsonNode);
+        if (!errors.isEmpty())
+        {
+            softAssert.recordFailedAssertion(
+                    String.format("JSON is not valid against schema: %s", StringUtils.join(errors, ", ")));
```
Comment (valfirst) [inline vividus-plugin-json/src/main/java/org/vividus/json/steps/JsonSteps.java:508]: how does error look like?

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-plugin-json/src/main/java/org/vividus/json/steps/JsonSteps.java [hunk 4]
```diff
@@ -485,6 +490,25 @@ public void assertNumberOfJsonElements(String json, String jsonPath, ComparisonR
         assertJsonElementsNumber(jsonPath, actualNumber, comparisonRule, elementsNumber);
     }
 
+    /**
+     * Validates json against json schema
```
Comment (ikalinin1) [inline vividus-plugin-json/src/main/java/org/vividus/json/steps/JsonSteps.java:494]: I believe javadoc should mention the version of the schema

File: docs/modules/plugins/pages/plugin-json.adoc [hunk 5]
```diff
@@ -500,3 +502,59 @@ Then number of JSON elements from `
 ]
 ` by JSON path `$..accountId` is equal to 2
 ----
+
+=== Validate JSON document
```
Comment (ikalinin1) [inline docs/modules/plugins/pages/plugin-json.adoc:506]:
```suggestion
=== Validate JSON against schema
```

File: vividus-plugin-json/src/main/java/org/vividus/json/steps/JsonSteps.java (hunk 3 above)
Comment (ikalinin1) [inline vividus-plugin-json/src/main/java/org/vividus/json/steps/JsonSteps.java:508]: java.lang.String#join(java.lang.CharSequence, java.lang.Iterable<? extends java.lang.CharSequence>) ?

File: vividus-plugin-json/src/main/java/org/vividus/json/steps/JsonSteps.java (hunk 2 above)
Comment (ikalinin1) [reply vividus-plugin-json/src/main/java/org/vividus/json/steps/JsonSteps.java:501]: @valfirst  could you please elaborate

Comment (uarlouski) [reply vividus-plugin-json/src/main/java/org/vividus/json/steps/JsonSteps.java:501]: @ikalinin1 the schema itself specfies its version "$schema": "https://json-schema.org/draft/2020-12/schema",

File: vividus-plugin-json/build.gradle [hunk 6]
```diff
@@ -13,6 +13,7 @@ dependencies {
     implementation(group: 'io.qameta.allure', name: 'allure-jsonunit', version: versions.allure) {
         exclude (group: 'io.qameta.allure')
     }
+    implementation(group: "com.networknt", name: "json-schema-validator", version: "1.0.76");
```
Comment (uarlouski) [inline vividus-plugin-json/build.gradle:16]: singlequotes as per all other cases where string interpolation doesn't happen

File: vividus-tests/src/main/resources/story/integration/JSON.story [hunk 7]
```diff
@@ -185,3 +185,44 @@ Meta:
...
+    },
+    "price": {
+      "type": "number",
+      "exclusiveMinimum": 0
+    },
+    "tags": {
+      "description": "Tags for the product",
+      "type": "array",
+      "items": {
+        "type": "string"
```
Comment (ikalinin1) [inline vividus-tests/src/main/resources/story/integration/JSON.story:220]: please update this example to validate that the array contains a specific list of items  not any strigns

File: vividus-plugin-json/src/main/java/org/vividus/json/steps/JsonSteps.java [hunk 8]
```diff
@@ -485,6 +493,45 @@ public void assertNumberOfJsonElements(String json, String jsonPath, ComparisonR
...
```
Comment (valfirst) [inline vividus-plugin-json/src/main/java/org/vividus/json/steps/JsonSteps.java:497]: make javadoc as much possible close to the documentation

File: vividus-plugin-json/src/main/java/org/vividus/json/steps/JsonSteps.java [hunk 9]
```diff
@@ -485,6 +493,45 @@ public void assertNumberOfJsonElements(String json, String jsonPath, ComparisonR
...
+     * @param json JSON to validate
+     * @param schema JSON schema with or without given schema specification. If schema specification is absent, JSON
+     *               is validated against `2020-12` specification
+     */
+    @Then("JSON `$json` is valid against schema `$schema`")
+    public void validateJsonAgainstSchema(String json, String schema)
+    {
+        JsonNode schemaNode = jsonUtils.readTree(schema);
+        SpecVersion.VersionFlag version;
+        if (schemaNode.has(SCHEMA_TAG))
```
Comment (valfirst) [inline vividus-plugin-json/src/main/java/org/vividus/json/steps/JsonSteps.java:507]: having this complexity added it makes sense to move the step to a new class

File: vividus-plugin-json/src/main/java/org/vividus/json/steps/JsonSteps.java [hunk 10]
```diff
@@ -485,6 +493,45 @@ public void assertNumberOfJsonElements(String json, String jsonPath, ComparisonR
...
+        }
+        else
+        {
+            version = SpecVersion.VersionFlag.V202012;
+        }
+        JsonSchemaFactory factory = JsonSchemaFactory.getInstance(version);
+        JsonSchema jsonSchema = factory.getSchema(schemaNode);
+        JsonNode jsonNode = jsonUtils.readTree(json);
+        Set<ValidationMessage> errors = jsonSchema.validate(jsonNode);
+        if (!errors.isEmpty())
```
Comment (valfirst) [inline vividus-plugin-json/src/main/java/org/vividus/json/steps/JsonSteps.java:528]: we need to record passed assertion in case of success

File: vividus-plugin-json/src/main/java/org/vividus/json/steps/JsonSteps.java [hunk 11]
```diff
@@ -485,6 +493,45 @@ public void assertNumberOfJsonElements(String json, String jsonPath, ComparisonR
...
+        {
+            softAssert.recordFailedAssertion("JSON is not valid against schema");
+            errors.forEach(error -> softAssert.recordFailedAssertion(error.toString()));
```
Comment (valfirst) [inline vividus-plugin-json/src/main/java/org/vividus/json/steps/JsonSteps.java:531]: there should be a single assertion error with all schema violations listed

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-plugin-json/src/main/java/org/vividus/json/steps/JsonValidationSteps.java [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+    private String formatValidationMessages(Set<ValidationMessage> validationMessages)
+    {
+        List<String> messages = new ArrayList<>();
+        int index = 1;
+        for (ValidationMessage validationMessage : validationMessages)
+        {
+            messages.add(String.format("%d) %s", index, validationMessage.toString()));
+            index++;
+        }
+        return String.join(String.format(",%n"), messages);
```
Comment (valfirst) [inline vividus-plugin-json/src/main/java/org/vividus/json/steps/JsonValidationSteps.java:107]: why not to use `StringBuilder`?

File: vividus-plugin-json/src/main/resources/vividus-plugin/spring.xml [hunk 13]
```diff
@@ -12,16 +12,17 @@
 
     <bean id="json-context" class="org.vividus.json.variable.JsonContextVariable" />
 
-    <bean id="jsonSteps" class="org.vividus.json.steps.JsonSteps">
-        <constructor-arg>
-            <bean class="org.vividus.json.softassert.JsonSoftAssert" parent="softAssert" />
-        </constructor-arg>
-    </bean>
```
Comment (valfirst) [inline vividus-plugin-json/src/main/resources/vividus-plugin/spring.xml:19]: why is this change needed?

Comment (IBrun) [reply vividus-plugin-json/src/main/resources/vividus-plugin/spring.xml:19]: To move definition of JsonSoftAssert bean from definition of jsonSteps bean

**Review by valfirst — CHANGES_REQUESTED** [review]

**Review by valfirst — APPROVED** [review]

**Review by uarlouski — APPROVED** [review]

**Review by ikalinin1 — APPROVED** [review]
